compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.ir.IrArrayBuilder
+import org.jetbrains.kotlin.backend.jvm.ir.createJvmIrBuilder
+import org.jetbrains.kotlin.backend.jvm.ir.irArray
+import org.jetbrains.kotlin.backend.jvm.ir.irArrayOf
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.builtins.PrimitiveType
+import org.jetbrains.kotlin.builtins.UnsignedType
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.declarations.IrFunction
+import org.jetbrains.kotlin.ir.declarations.IrPackageFragment
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.symbols.IrFunctionSymbol
+import org.jetbrains.kotlin.ir.util.*
+
+val varargPhase = makeIrFilePhase(
+    ::VarargLowering,
+    name = "VarargLowering",
+    description = "Replace varargs with array arguments and lower arrayOf calls"
+)
+
+private class VarargLowering(val context: JvmBackendContext) : FileLoweringPass, IrElementTransformerVoidWithContext() {
+    override fun lower(irFile: IrFile) = irFile.transformChildrenVoid()
+
+    // Ignore annotations
+    override fun visitConstructorCall(expression: IrConstructorCall): IrExpression {
+        val constructor = expression.symbol.owner
+        if (constructor.constructedClass.isAnnotationClass)
+            return expression
+        return super.visitConstructorCall(expression)
+    }
+
+    override fun visitFunctionAccess(expression: IrFunctionAccessExpression): IrExpression {
+        expression.transformChildrenVoid()
+        val function = expression.symbol
+
+        // Replace empty varargs with empty arrays
+        for (i in 0 until expression.valueArgumentsCount) {
+            if (expression.getValueArgument(i) != null)
+                continue
+
+            val parameter = function.owner.valueParameters[i]
+            if (parameter.varargElementType != null && !parameter.hasDefaultValue()) {
+                val arrayType = parameter.type.substitute(expression.typeSubstitutionMap)
+                expression.putValueArgument(i, createBuilder().irArrayOf(arrayType))
+            }
+        }
+
+        // Lower arrayOf calls
+        return if (function.isArrayOf) expression.getValueArgument(0)!! else expression
```

Comment:
It's actually safe, since the call to `function.isArrayOf` ensures that the function in question only has a single vararg parameter and the code above ensure that this parameter is not null. I've added a comment to make it less confusing.